internal/selfupdate/download.go

```diff
@@ -74,33 +66,47 @@ func extractToFile(buf []byte, filename, target string, printf func(string, ...i
 		rd = file
 	}
 
-	err = os.Remove(target)
-	if os.IsNotExist(err) {
-		err = nil
-	}
+	// Write everything to a temp file
+	new, err := os.CreateTemp("", "restic")
 	if err != nil {
-		return fmt.Errorf("unable to remove target file: %v", err)
+		return err
 	}
 
-	dest, err := os.OpenFile(target, os.O_CREATE|os.O_EXCL|os.O_WRONLY, mode)
+	n, err := io.Copy(new, rd)
```

Comment:
I suggest Sync'ing the file before trying to move it in place. This only works when it's in the same dir as the target, since moving across mount points means copy-then-remove.

```suggestion
	n, err := io.Copy(new, rd)
	if err == nil {
		err = new.Sync()
	}
```

Reply:
Done